### ISSUE TITLE
Restrict values that trigger "Convert to Utf-8 string" analyzer

### ISSUE THREAD

**Issue by davidwengier** [dotnet/roslyn#62207]
At the moment we will happily suggest any byte array be converted to a UTF string literal, if the values would make for valid UTF-8 strings. We should be a little more strict about it, to avoid some false positives. The following examples all come from the runtime repo.

eg: None of these are characters
https://github.com/dotnet/runtime/blob/6a889d234267a4c96ed21d0e1660dce787d78a38/src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Conversion.cs#L114-L126

https://github.com/dotnet/runtime/blob/b5a8ece073110140e2d9696cdfdc047ec78c2fa1/src/libraries/System.Private.CoreLib/src/System/DateTime.cs#L117-L118

**Comment by CyrusNajmabadi** [comment]
Oooh.  please show me what those get converted to :D

**Comment by davidwengier** [comment]
@stephentoub might have screenshots, but otherwise I'll be sure to add some when I get to this :)

**Comment by stephentoub** [comment]
```C#
private static readonly byte[][] s_simpleTypeConversions =
{
    // to:                   BYTE I2   I4   I8   FLT  DBL  DEC  CHAR BOOL SBYTE U2   U4   U8
    /* from */
        "B"u8.ToArray(),
        "B"u8.ToArray(),
        "B"u8.ToArray(),
        "B"u8.ToArray(),
        "C"u8.ToArray(),
        "C"u8.ToArray(),
        "CCCCCCCCCCC"u8.ToArray(),
        "B"u8.ToArray(),
        ""u8.ToArray(),
        "B"u8.ToArray(),
        "B"u8.ToArray(),
        "B"u8.ToArray(),
        "B"u8.ToArray(),
};
```
and
```C#
private static ReadOnlySpan<byte> DaysInMonth365 => ""u8.ToArray();
private static ReadOnlySpan<byte> DaysInMonth366 => ""u8.ToArray();
```
This also highlights another issue... those `.ToArray()`s shouldn't be there in the latter case.

**Comment by davidwengier** [comment]
Looks beautiful. You have something against ASCII art?! 😛

**Comment by CyrusNajmabadi** [comment]
I agree with david.  That's genuinely amazing :D

**Comment by davidwengier** [comment]
All jokes aside, I am honestly not so sure about this any more. In the `DaysInMonth365` case, its pretty clear that leaving those as numbers makes the most sense, and communicates best what the data means, but I'm not entirely convinced that there isn't someone who has `new byte[] _listOfFancyASCIICharacters = ...` that wouldn't prefer the UTF8 string format.

**Comment by Youssef1313** [comment]
> This also highlights another issue... those `.ToArray()`s shouldn't be there in the latter case.

Yeah I pointed to that in https://github.com/dotnet/roslyn/pull/61650#pullrequestreview-993136590.

@davidwengier Will you track both issues in this single one?

**Comment by stephentoub** [comment]
> I'm not entirely convinced that there isn't someone who has new byte[] _listOfFancyASCIICharacters = ... that wouldn't prefer the UTF8 string format.

I think the analyzer should err on the side of false negatives. It only helps with maintainability, and getting it wrong degrades maintainability. I suggest:
- If the values are all const chars cast to byte, e.g. `new byte[] { (byte)'h', (byte)'i'}`, convert it regardless of their values. 
- Otherwise, only convert if it contains a blessed subset of values, e.g. Unicode word and digit characters, \r, \n, \t, and a limited set of punctuation. dotnet/runtime now has several hundred uses of u8 in product source and several hundred more in tests... you could experiment with a heuristic to see how many of those would have been caught.

The goal isn't to catch every possible use, but to strike a good balance between false positives and false negatives.

>> This also highlights another issue... those .ToArray()s shouldn't be there in the latter case.

> Yeah I pointed to that in https://github.com/dotnet/roslyn/pull/61650#pullrequestreview-993136590.

Note that this is actually making the "fixed" code slower, taking something that wasn't actually allocating (even though it looks like it is) and forcing it to allocate, so fixing it should be a priority, more so than the aforementioned false positives.

**Comment by davidwengier** [comment]
> I think the analyzer should err on the side of false negatives

Yeah, you're probably right. This is an interesting case:

https://github.com/dotnet/runtime/blob/main/src/libraries/Common/tests/Tests/System/Net/aspnetcore/Http3/QPackDecoderTest.cs#L29
`new byte[] { 0x37, 0x02, 0x74, 0x72, 0x61, 0x6e, 0x73, 0x6c, 0x61, 0x74, 0x65 };` is `"7translate"u8`, so _most_ of the string does actually make sense 🤷‍♂️

**Comment by davidwengier** [comment]
Thanks @Youssef1313, I thought I had a note for following that up, but must have forgot. Will just fix both issues with this.
